### PR TITLE
KAFKA-17414: Move RequestLocal to server-common module

### PR DESCRIPTION
**Pull request by wernerdv**
Moved RequestLocal from core to server-common module.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by wernerdv** [comment]
@chia7712 Hello, PTAL.

**Comment by chia7712** [comment]
@wernerdv please fix the conflicts, thanks!

**Comment by wernerdv** [comment]
@chia7712 conflicts are resolved.

File: server-common/src/main/java/org/apache/kafka/server/common/RequestLocal.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    /** The returned instance should be confined to a single thread. */
+    public static RequestLocal withThreadConfinedCaching() {
+        return new RequestLocal(BufferSupplier.create());
+    }
+
+    public BufferSupplier bufferSupplier() {
+        return bufferSupplier;
+    }
+
+    public void close() {
```
Comment (chia7712) [inline server-common/src/main/java/org/apache/kafka/server/common/RequestLocal.java:51]: Could you please make `RequestLocal` extend `AutoCloseable`?

File: server-common/src/main/java/org/apache/kafka/server/common/RequestLocal.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    public static final RequestLocal NO_CACHING = new RequestLocal(BufferSupplier.NO_CACHING);
+
+    private final BufferSupplier bufferSupplier;
+
+    public RequestLocal(BufferSupplier bufferSupplier) {
+        this.bufferSupplier = bufferSupplier;
+    }
+
+    /** The returned instance should be confined to a single thread. */
+    public static RequestLocal withThreadConfinedCaching() {
```
Comment (chia7712) [inline server-common/src/main/java/org/apache/kafka/server/common/RequestLocal.java:42]: Could you please add static method `noCaching` for consistency? the method can return `NO_CACHING` directly. Also, `NO_CACHING` can be private variable

**Review by chia7712 — COMMENTED** [review]
@wernerdv thanks for your patch

File: server-common/src/main/java/org/apache/kafka/server/common/RequestLocal.java (hunk 1 above)
Comment (wernerdv) [reply server-common/src/main/java/org/apache/kafka/server/common/RequestLocal.java:51]: Done

File: server-common/src/main/java/org/apache/kafka/server/common/RequestLocal.java (hunk 2 above)
Comment (wernerdv) [reply server-common/src/main/java/org/apache/kafka/server/common/RequestLocal.java:42]: Done

**Review by chia7712 — APPROVED** [review]

**Pull request merged**